Quantum Models

DanPuzzuoli opened the pull request:
### Summary

This PR introduces classes for representing quantum models:
- `HamiltonianModel` - a time-dependent Hamiltonian subclassing `OperatorModel`
- `QuantumSystemModel` - a class representing a quantum system, containing a `HamiltonianModel`, as well as an optional list of noise operators and noise signals.

We will likely want to expand the above classes to offer much more functionality, but for now they are minimally defined to support constructing/solving differential equations (described in the details section).

For integrating with the differential equation solving, we define new problems:
- `SchrodingerProblem` - sets up the Schrodinger equation
- `LindbladProblem` - sets up the Lindblad equation for a density matrix
- `LindbladianProblem` - sets up the Lindblad equation for computing the Lindbladian super operator

The problem interfaces will need further design/work, but for now this PR at least gets some functioning pieces.

This PR also adds a small file `pulse_new/de/solve.py` containing a higher level `solve` function that solves a fully specified `BMDE_Problem` (i.e. one specified with a full time interval instead of just an initial time). This is to simplify solving when a single solver call is all that is needed (i.e. combining the solver construction, integration call, and state request `solver.y` into one line).


### Details

Here we describe in detail what this PR does

Introduction of quantum model classes in `pulse_new/models/quantum_models.py`:
- `HamiltonianModel(OperatorModel)`
   - Only modified functionality is: `__init__` checks of operators are Hermitian, and the `evaluate` method has modified structure due to the meaning of a "frame" for a Hamiltonian. I.e. for us frame operators are anti-Hermitian, i.e. F=-iH0, and the default frame handling is to evaluate exp(-tF)G(t)exp(tF)-F, where G(t) is the generator, which for a hamiltonian would be G(t)=-iH(t). For a Hamiltonian evaluated in a frame F=-iH, we actually want to compute exp(-tF)H(t)exp(tF)-H0, so we modify the `evaluate` function to return exp(-tF)H(t)exp(tF) - iF (as everything is done in terms of F instead of H0).
- `QuantumSystemModel`
   - Contains a `HamiltonianModel`, and list of noise operators and noise signals. As mentioned in the summary, the only functionality this class has is a single property, `vectorized_lindblad_generator`, which constructs and returns an `OperatorModel` object corresponding to the vectorized Lindblad equation in column stacking convention.

Introduction of new DE problems in `pulse_new/de/DE_Problems.py`
- `SchrodingerProblem(BMDE_Problem)`
    - Constructs the Schrodinger generator as an `OperatorModel` from a `HamiltonionModel`.
    - We can refine the behaviour later but for now this does no type checking on the state, as this structure works for either unitaries or state vectors.
- `LindbladProblem(BMDE_Problem)` - Lindblad equation for density matrices
   - Constructs the vectorized Lindblad generator as an `OperatorModel` from a `QuantumSystemModel`
   - Does no type checking, but assumes the state is a 2d array. To handle states, constructs a `StateTypeConverter` that specifies the inner type as a 1d array. The effect of this is: the subsequently constructed solver will now internally represent the state as a 1d array (the vectorized density matrix), but when interacting with the user - e.g. when returning the state to the user - will automatically reshape things to the original 2d shape, so that the user need not ever know that vectorization is occuring.
- `LindbladianProblem(BMDE_Problem)` - Lindblad equation for density matrices
   - Exactly the same as `LindbladProblem`, but just assumes the state is already given as a 2d array representing the initial Lindbladian super operator.

Added file for high level solver interface `pulse_new/de/solve.py`
- Contains function `solve`, for single-line solving of a BMDE problem.

Changes to `pulse_new/de/type_utils.py`
- Moved from `de` folder into `pulse_new`
- Modified `StateTypeConverter` to enable specification of the `order` keyword argument when calling array reshaping methods. This was to enable column-stacking convention vectorization, which is the convention used in `terra.quantum_info` (the default in `numpy` is row-stacking). I've now set the default order to `'F'`, which corresponds to column stacking.
- Also put some helper functions for vectorizing the Lindblad equation into here as they needed to be used in multiple places, and are naturally connected to the array reshaping functionality of `StateTypeConverter`.

Added tests. The tests are a lot of the lines in this PR but they are all basically copied from existing tests (as all new functionality comes from subclasses) with minor modifications.
- `models/test_hamiltonian_model.py` - testing specialized evaluation and frame handling.
- `models/test_quantum_system_model.py` - testing the output of `vectorized_lindblad_generator`. Ensuring that the vectorized evaluation is consistent with an independent, non-vectorized computation of the RHS of the Lindblad equation.
- `de/test_de_problems.py` - added tests for `SchrodingerProblem`, `LindbladProblem`, and `LindbladianProblem`
- `test_type_utils.py` (moved from `de/test_type_utils.py`) - modified existing tests to work with array reshaping `order='C'`, which is the default in `numpy`, and added new tests to work with `order='F'`

## Original text below

### ~Overview~

~Add "quantum models":~
~1. `HamiltonianModel` - representation of a time-dependent Hamiltonian.~
~2. `QuantumSystemModel` - containing a `HamiltonianModel` as well as a noise model (Lindblad operators, potentially with signal coefficients).~

~Add relevant DE problems:~
~1. `SchrodingerProblem` - setup the Schrodinger equation with a `HamiltonianModel`.~
~2. `LindbladProblem` - setup the vectorized Lindblad master equation with a `QuantumSystemModel`.~

~For the initial states with the DE problems:~
~- I think it makes sense to allow the user to specify an arbitrary array of the appropriate shape, however:~
~- These problems should recognize `quantum_info` types and behave appropriately, e.g. `Statevector`, `DensityMatrix`, `Unitary`, `Operator`, `SuperOp` (or any other object that can be converted to this). The constructed solver should report states in the relevant type (while still working internally just with arrays).~

### ~Summary~

~For this PR I'm not sure if this handling of state types should be left for later. Perhaps for now we can just make things work with arrays, verify the correctness of computations, and leave the type handling for subsequent work (will likely require a lot of work/design).~

~For now:~
~- A `HamiltonianModel(OperatorModel)` has been introduced, along with a `SchrodingerProblem` class.~
~- A `QuantumSystemModel` class, along with a `LindbladProblem` class, has been introduced.~

~To do:~
~- The vectorized Lindblad equation needs to be changed to column stacking convention to be consistent with `SuperOp`. This requires handling row v.s. column stacking in `type_utils.py`.~

DanPuzzuoli commented:
~Redid the details of this in light of recent changes to `OperatorModel` and surrounding classes.~

~Going to add a subclasses of BMDE problem for the Schrodinger equation, utilizing the new `HamiltonianModel`.~

This comment is very out of date.

DanPuzzuoli commented:
Note: just added an "order" attribute to `StateTypeConverter` for array reshaping commands. (Purpose is to enable vectorization in column stacking convention, but the default 2d->1d array reshaping is row stacking.)

Set the default to be `order='F'`, corresponding to column stacking convention in the 2d<->1d case, so this argument should ideally not need to be used.

chriseclectic commented:
Can you rename to:

chriseclectic commented:
Still haven't had a change to go through all this yet, but a couple of superficial comments regarding names. Can we change:
`QuantumSystemModel` -> `LindbladModel` (or maybe `LinbladianModel`?)
`SchrodingerProblem` -> `StatevectorProblem`
`LindbladProblem` -> `DensityMatrixProblem`
`LindbladianProblem` -> `SuperOpProblem` (or maybe `QuantumChannelProblem`)
and add a `UnitaryProblem`

chriseclectic commented on qiskit/providers/aer/pulse_new/de/DE_Problems.py:
```diff
@@ -127,6 +129,190 @@ def __init__(self,
         if cutoff_freq is not None:
             self._generator.cutoff_freq = cutoff_freq
 
+class SchrodingerProblem(BMDE_Problem):
```
```suggestion
class StatevectorProblem(BMDE_Problem):
```

DanPuzzuoli replied:
Kept `SchrodingerProblem`, but subclassed it with a `StatevectorProblem` and a `UnitaryProblem`. I think we can think of `SchrodingerProblem` as a generic DE constructor that doesn't really worry about type, whereas `StatevectorProblem` is really meant to be for a vector, and `UnitaryProblem` is really meant to be for a unitary. E.g. The `UnitaryProblem` will now set `y0` to be the identity matrix if none is specified. 

I haven't put any specialized logic into `StatevectorProblem` (e.g. checking that the state is actually a vector, or giving a default state), but I think at this point given the future structural changes to the problems/solvers we should save this for later.

chriseclectic commented on qiskit/providers/aer/pulse_new/de/DE_Problems.py:
```diff
@@ -127,6 +129,190 @@ def __init__(self,
             self._generator.cutoff_freq = cutoff_freq
 
+class SchrodingerProblem(BMDE_Problem):
+    """A Schrodinger equation problem, where the generator is of the form
```
Maybe something like "A differential equation problem for statevector evolution according to the Schrodinger equation...."""

DanPuzzuoli replied:
Done.

chriseclectic commented on qiskit/providers/aer/pulse_new/de/DE_Problems.py:
```diff
@@ -127,6 +129,190 @@ def __init__(self,
+
+    def __init__(self,
+                 hamiltonian: HamiltonianModel,
+                 y0: Optional[np.ndarray] = None,
```
I think we can move `y0`, `t0`, `interval` out of the problem and make them arguments of the `solve` equation.

DanPuzzuoli replied:
Resolving this as it's part of the future issue/PR summary.

chriseclectic commented on qiskit/providers/aer/pulse_new/de/DE_Problems.py:
```diff
@@ -127,6 +129,190 @@ def __init__(self,
+                         frame=frame,
+                         cutoff_freq=cutoff_freq)
+
+class LindbladProblem(BMDE_Problem):
```
```suggestion
class DensityMatrixProblem(BMDE_Problem):
```

chriseclectic commented on qiskit/providers/aer/pulse_new/de/DE_Problems.py:
```diff
@@ -127,6 +129,190 @@ def __init__(self,
+                         cutoff_freq=cutoff_freq)
+
+class LindbladProblem(BMDE_Problem):
+    """Simulate a density matrix according to the Lindblad equation:
```
```suggestion
    """Simulate density matrix evolution according to the Lindblad equation:
```

chriseclectic commented on qiskit/providers/aer/pulse_new/de/DE_Problems.py:
```diff
@@ -127,6 +129,190 @@ def __init__(self,
+
+            # if a frame was additionally specified as an argument, warn
+            if frame != 'auto':
+                warn("""A frame was specified in both the QuantumSystemModel
```
Does this need to be a warning? Is the idea of allowing frames in both the Hamiltonian and problem classes that I could define a Hamiltonian in a certain frame, and then solve in a different frame? If this is the case can the frame transformations just be combined so that you go out of one frame and into the other?

DanPuzzuoli replied:
Yeah you basically summarized it: the point is to provide a mechanism for the following use cases:
- The user wants to work in the lab frame, but simulate behind the scenes, for numerical purposes, in a different frame, potentially with cutoff frequencies. In this case, the original model the user holds shouldn't be modified by these transformations, and results should be returned in the lab frame.
- The user is fully working in a frame (i.e. the model has a frame), and they want to solve and get results in that frame. In this case they can define the problem using a model that is already in the frame of interest. The problem detects this, and then sets an internal flag to simulate in that frame and also return results in that frame. 

The warning is to prevent collision of these.

This functionality is important but I do feel like this implementation is slightly awkward. The problems and this logic were written before the `Frame` objects, so maybe using those this could be simplified (maybe in a subsequent PR). Perhaps we can change the problem arguments to be something like: 
- change the name of the problem `frame` argument to `solve_in_frame` (making explicit that this is a frame of the solver), and 
- add a `results_in_frame` kwarg, specifying which frame to return results in (which would default to the frame the model is in).
- similarly, we could change `cutoff_freq` to `solve_with_cutoff_freq` or something like this, to again make explicit that this is purely a property of the simulation, not of the model object the user is working with

with this I don't think there is any need anymore for a warning.

chriseclectic commented on qiskit/providers/aer/pulse_new/de/DE_Problems.py:
```diff
@@ -127,6 +129,190 @@ def __init__(self,
+        # handle frame
+        # if frame is in hamiltonian, set the generator frame to the vectorized
+        # version of it
+        if q_model.hamiltonian.frame.frame_operator is not None:
```
Do we want to allow two locations to set frames? Or can the problem just update the frame of the Hamiltonian or vice versa?

DanPuzzuoli replied:
Going to resolve this as it's discussed in my previous response.

chriseclectic commented on qiskit/providers/aer/pulse_new/de/DE_Problems.py:
```diff
@@ -127,6 +129,190 @@ def __init__(self,
+                         cutoff_freq=cutoff_freq,
+                         state_type_converter=converter)
+
+class LindbladianProblem(BMDE_Problem):
```
```suggestion
class SuperOpProblem(BMDE_Problem):
```

chriseclectic commented on qiskit/providers/aer/pulse_new/de/DE_Problems.py:
```diff
@@ -127,6 +129,190 @@ def __init__(self,
+class LindbladianProblem(BMDE_Problem):
+    """Define a Lindbladian simulation problem.
+
+    I.e. Construct the Lindblad equation as in :class:`LindbladProblem`,
```
"Simulate superoperator evolution according to the Lindlad equation..."

DanPuzzuoli replied:
Changed.

chriseclectic commented on qiskit/providers/aer/pulse_new/de/DE_Problems.py:
```diff
@@ -127,6 +129,190 @@ def __init__(self,
+        # handle frame
+        # if frame is in hamiltonian, set the generator frame to the vectorized
+        # version of it
+        if q_model.hamiltonian.frame.frame_operator is not None:
```
Same as above. Also could this frame changing functionality be moved to a method like `update_frame` or `set_frame`?

DanPuzzuoli replied:
See comment about frames and problems - will resolve this one as well.

chriseclectic commented on qiskit/providers/aer/pulse_new/de/DE_Solvers.py:
```diff
@@ -18,7 +18,7 @@
-from .type_utils import StateTypeConverter
+from ..type_utils import StateTypeConverter
 
 class BMDE_Solver:
```
To continue our previous discussion I don't think this class really needs to exist. It's functionality could be split between the `BMDE_Problem` class and the `solve` function.

DanPuzzuoli replied:
See the discussion in a different comment about `StateTypeConverter`. Will resolve this here and discussion can be had elsewhere.

chriseclectic commented on qiskit/providers/aer/pulse_new/de/solve.py:
```diff
@@ -0,0 +1,46 @@
+from .DE_Options import DE_Options
+from .DE_Solvers import BMDE_Solver
+
+def solve(problem : BMDE_Problem, options: Optional[DE_Options] = None):
```
This function could be made to be closer in signature to the `solve_ivp` function from scipy. Something like

```python
solve_ivp(rhs: BMDE_Problem, t_span: Tuple[float, float], y0: array_like, method: Solver, options)
```

DanPuzzuoli replied:
Resolving this as it's part of the future issue/PR summary.

chriseclectic commented on qiskit/providers/aer/pulse_new/de/solve.py:
```diff
@@ -0,0 +1,46 @@
+        options = DE_Options()
+
+    # construct solver, integrate to end of interval, return result
+    solver = BMDE_Solver(problem, options)
```
The functionality of the BMDE_solver (frame transformations and integerate calls) can be moved here.

DanPuzzuoli replied:
Resolving this as it's part of the future issue/PR summary.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/frame.py:
```diff
@@ -430,17 +430,21 @@ def __init__(self, frame_operator: Union[Operator, np.array]):
         elif isinstance(frame_operator, np.ndarray) and frame_operator.ndim == 1:
 
+            # first check if it is Hermitian, if so convert to anti-Hermitian
+            if np.linalg.norm(frame_operator - frame_operator.conj()) < 1e-10:
```
Is `norm` required here over just doing `np.allclose(frame_operator, frame_operator.conj(), atol=atol, rtol=rtol)`? We should also add the threshold value as an `atol` kwarg or variable.

DanPuzzuoli replied:
Changed.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/frame.py:
```diff
@@ -430,17 +430,21 @@ def __init__(self, frame_operator: Union[Operator, np.array]):
+                frame_operator = -1j * frame_operator
+
             # verify that it is anti-hermitian (i.e. purely imaginary)
             if np.linalg.norm(frame_operator + frame_operator.conj()) > 1e-10:
```
similarly here but with `np.allclose(frame_operator, -frame_operator.conj()`

DanPuzzuoli replied:
Done.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/frame.py:
```diff
@@ -452,10 +456,14 @@ def __init__(self, frame_operator: Union[Operator, np.array]):
             frame_operator = Operator(frame_operator)
 
+            # first check if it is Hermitian, if so convert to anti-Hermitian
+            if np.linalg.norm((frame_operator - frame_operator.adjoint()).data) < 1e-10 :
```
There is an `is_hermitian_matrix` predicate in `qiskit.quantum_info.operators.predicates` you can use here (we could also add an `is_antihermitian_matrix` one)

DanPuzzuoli replied:
Nice, switched to using this.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/frame.py:
```diff
@@ -452,10 +456,14 @@ def __init__(self, frame_operator: Union[Operator, np.array]):
+            if np.linalg.norm((frame_operator - frame_operator.adjoint()).data) < 1e-10 :
+                frame_operator = -1j * frame_operator
+
             # verify anti-hermitian
```
This should be in an `else` block, since if the input was hermitian we know that `-1j * mat` is anti-hermitian so we don't need to check again.

chriseclectic replied:
We should also put the order of the checks (check hermitian then anti, or vice versa) based on the standard way we want to encourage inputing frames so that the least number of checks need to happen for standard use cases

DanPuzzuoli replied:
Changed to an if/else block.

In terms of check order, I suspect users will most often specify it with a Hermitian operator, so I'll leave this as the first check.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/frame.py:
```diff
@@ -452,10 +456,14 @@ def __init__(self, frame_operator: Union[Operator, np.array]):
+
             # verify anti-hermitian
             herm_part = frame_operator + frame_operator.adjoint()
             if herm_part != Operator(np.zeros(frame_operator.dim)):
```
This would be more efficient to check as:
```
np.allclose(frame_operator.data, -frame_operator.adjoint().data, atol, rtol)
```

DanPuzzuoli replied:
Done.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/quantum_models.py:
```diff
@@ -0,0 +1,245 @@
+            if isinstance(operator, Operator):
+                operator = operator.data
+
+            if np.linalg.norm((operator.conj().transpose()
```
You can use the `is_hermitian_matrix` predicate

DanPuzzuoli replied:
Done.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/quantum_models.py:
```diff
@@ -0,0 +1,245 @@
+    :class:`OperatorModel`, with the following modifications:
+        - The operators in the linear decomposition are verified to be
+          Hermitian.
+        - Frames are dealt with assuming the structure of the Schrodinger
```
Maybe the base class should have a `update_frame(frame_op)` or `set_frame` method that can be overloaded in the subclasses to make this clearer

chriseclectic replied:
Actually after looking at notebook I guess this is already done with a setter `ham.frame = frame`

chriseclectic commented on qiskit/providers/aer/pulse_new/models/quantum_models.py:
```diff
@@ -0,0 +1,245 @@
+                         frame=frame,
+                         cutoff_freq=cutoff_freq)
+
+    def evaluate(self, time: float, in_frame_basis: bool = False) -> np.array:
```
Should the frame setting be done separate to this function. Something like:
```
model.set_frame(frame)
model.evaluate(t)
```
rather than `model.evaluate(t, in_frame_basis=True)`

DanPuzzuoli replied:
This is different than setting a frame. The "frame basis" throughout this code is defined as the basis in which the frame operator is diagonal (according to the output of `np.linalg.eigh`) - so setting `in_frame_basis=True` has no effect on which frame the evaluation occurs in, just the basis in which the matrix is represented.

This argument is purely for computational purposes - when solving in a rotating frame it makes sense to solve the problem in the basis in which the frame operator is diagonal to save on the cost of compute `expm(t * F)` for every `t` the solver queries.

I'll leave this unresolved in case you want to comment further.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/quantum_models.py:
```diff
@@ -0,0 +1,245 @@
+                                             operator_in_frame_basis=True,
+                                             return_in_frame_basis=in_frame_basis)
+
+class QuantumSystemModel:
```
I think there should be a common base class for all these `Model` classes that specifies the abstract interface like `evaluate` and applying frame transformations.

chriseclectic replied:
This should probably also be called `LindbladModel` to match `HamiltonianModel`

DanPuzzuoli replied:
Yeah makes sense, I'll make a note of this as something to potentially do in other PRs.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/quantum_models.py:
```diff
@@ -0,0 +1,245 @@
+          :math:`j^{th}` Lindblad operator.
+    """
+
+    def __init__(self,
```
Rather than having to construct the Hamiltonian model can this have the same signature as Hamiltonian model init but with the extra kwargs for noise operators / signals?

You could add a `from_hamiltonian` static method which mimics this init call

DanPuzzuoli replied:
Changed `QuantumSystemModel` to `LindbladModel` and it is now a subclass of `OperatorModel`. It now directly represents the vectorized Lindblad equation.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/quantum_models.py:
```diff
@@ -0,0 +1,245 @@
+        self.noise_signals = noise_signals
+
+    @property
+    def vectorized_lindblad_generator(self):
```
I think this is how this class is constructed (as a subclass of OperatorModel), rather than just being a container which you then call this method to convert to an OperatorModel

DanPuzzuoli replied:
I was thinking the `QuantumSystemModel` may have more functionality beyond just representing the Lindblad equation - e.g. it could also construct the DE required for monte carlo simulation.

DanPuzzuoli replied:
Resolving this as the implications of this comment are addressed byt he previous one.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/quantum_models.py:
```diff
@@ -0,0 +1,245 @@
+                             signals=full_signals)
+
+
+def to_array(op: Union[Operator, np.array, List[Operator], List[np.array]]):
```
You could probably just replace this function with `Operator(op).data`

DanPuzzuoli replied:
See comment below.

chriseclectic commented on qiskit/providers/aer/pulse_new/models/quantum_models.py:
```diff
@@ -0,0 +1,245 @@
+
+
+def to_array(op: Union[Operator, np.array, List[Operator], List[np.array]]):
+    """Convert an operator, either specified as an `Operator` or an array
```
Actually I see below this is supposed to work with lists or ops, so this doc string might need changing.

It will also have different ndims in output depending on if the input is an op or list of ops (ndim = 2 or 3)

DanPuzzuoli replied:
Moved `to_array` into `type_utils.py`, and changed documentation to make the cases it's supposed to work with clear.

chriseclectic commented on qiskit/providers/aer/pulse_new/type_utils.py:
```diff
@@ -13,46 +13,63 @@
 import numpy as np
 
 
 class StateTypeConverter:
```
I'm not sure I understand the purpose of this class. It seems to only be used in the denisty matrix problem / lindblad problem class, so could its functionality just be moved into that class?

DanPuzzuoli replied:
This class is also used in `DE_Methods.py` - I initially made it to automate vectorization in those classes as well. E.g. `solve_ivp` in `scipy` requires 1d arrays - the `ScipyODE(ODE_Method)` class contains an attribute that indicates this, and it automatically flattens the state of the de and wraps the inputs/outputs of the rhs function to work with the flattened version.

I was imagining this class to have utility as well when we start specifying initial states in terms of `terra.quantum_info` types. I also figured things like conversion between real/complex arrays could go here if that is ever needed.

chriseclectic reviewed (commented):
Other than some rename it seems like there are a few things that are a bit more complicated than they might need to be. Mainly to do with the density matrix / lindblad problems, and frame transformations.

* Can the QuantumSystemModel be an OperatorModel subclass
* Does there need to be so many locations to specify frames? If so can they all be just wrappers of calls to update a single frame in the `OperatorModel` class?
* Can the DE_Options class be used as a validation for dict options so they are just passed in as a dict or kwargs to solve?

I have a couple longer comments that I'll type in a separate comment

chriseclectic commented:
Looking at the example notebooks removing the BMDE_Solver class would bring to interface to look something like:

4. Solve Shrodinger equation with a HamiltonianModel
```python
hamiltonian = HamiltonianModel(operators=operators, signals=signals)
problem = SchrodingerProblem(hamiltonian)

# Solve with default frame
y = solve(problem, [0., T], y0, options={'atol'=1e-10, 'rtol'=1e-10})

# Solving with no frame
problem.frame = None
y = solve(problem, [0., T], y0, options={'atol'=1e-10, 'rtol'=1e-10})
```

There also seems to be a lot of overlap in between using Model and  Problem classes (like frames / cutoffs etc), especially if we remove the solve params (t0, tf, y0) from problems. In all the example notebooks you do all the work with the model and only use the problem to feed into the solve function. Maybe we should think about combining these two classes, so you can just work with model and models could allow different cases depending on type/shape or y0? Something like
HamiltonianModel -> y0 could be statevector, density matrix, or unitary (operator).
LindbladModel -> y0 could be statevector/density matrix, unitary (operator), or quantum channel (superop)

```python
# Schrodinger
model = HamiltonianModel(operators, signals)
y0 = Statevector([1, 0]) # or just [1, 0]
y = solve(model, [0, T], y0)

# Heisenberg
y0 = Operator([0, 1], [1, 0]) # or just [[0, 1], [1, 0]]
y = solve(model, [0, T], y0)

# Lindblad
model = LindbladModel(operators, signals, noise_operators, noise_signals)
y0 = DensityMatrix([[1, 0], [0, 1])
y = solve(model, [0, T], y0)

# Superop
y0 = SuperOp(Operator([0, 1], [1, 0]))
y = solve(model, [0, T], y0)
```

And general ODEs would still be
```
model = OperatorModel(...)
y0 = np.array(...)
y = solve(model, [0, T], y0)
```

DanPuzzuoli commented:
Based on comments I've made some changes, replied to a few for further discussion, and have made notes on things to be done/addressed in future issues/PRs. If I either directly implemented the change you suggested, or classified the comment under the "future issues/PRs" category, I've replied and resolved the conversation, otherwise I've replied and left the comment up.

Due to the size of this PR and the fact that the functionality it aims to introduce is generally working I think it'd be good to merge this soon and handle the remaining points separately.

In conjunction with the changes here, I've also updated the `general_demo.ipynb` in `refactor/demos` to align with these changes.

Summary of changes:
- Various small technical changes as suggested, e.g. using `is_hermitian_matrix` and combining checks on Hermiticity/anti-Hermiticity into `if`/`else` blocks to limit the number of checks being made.
- Kept `SchrodingerProblem`, but created subclasses `StatevectorProblem` and `UnitaryProblem`. The idea here is that the generator construction and frame handling choices are shared in `StatevectorProblem` and `UnitaryProblem`, so those can be kept in `SchrodingerProblem`.
- Changed `LindbladProblem` to `DensityMatrixProblem` and `LindbladianProblem` to `SuperOpProblem`.
- Changed `QuantumSystemModel` to `LindbladModel`, and made it a subclass of `OperatorModel`. It now directly represents the vectorized Lindblad equation. Also added a `from_hamiltonian` constructor as suggested.

Summary of comments for future issues/PRs:
- Remove `BMDE_Solver` and make the interface of `solve` more like `solve_ivp` in `scipy`. Breaking it down roughly:
    - Move functionality in `BMDE_Solver` regarding input/output types into `solve`.
    - Rename `ODE_Method` and `BMDE_Method` to `ODE_Solver` and `BMDE_Solver` respectively.
    - Modify the interface to `solve` (along the structure of `ODE_Solver` and `BMDE_Solver`) to more closely resemble `solve_ivp` in `scipy`, potentially  also making `ODE_Solver` and `BMDE_Solver` resemble the corresponding `scipy` classes. E.g. move the `y0` and time interval arguments from the problem to `solve`.
- Modify how DE options passed to avoid having to instantiate the `DE_Options` class. E.g. move them as kwargs to solve, and the `DE_Options` class can maybe just be used behind the scenes to fill out defaults.
- Clarify frame handling in the `BMDE_Problem` interface
    - Change `frame` keyword arg to `solve_in_frame` to make it clear that this frame specification is purely for the solver
    - Similarly, change the kwarg `cutoff_freq` to `solve_with_cutoff_freq`
    - Add a `results_in_frame` kwarg to specify the frame that the user wants results in
- Add abstract base classes for `HamiltonianModel` and `QuantumSystemModel` (or `LindbladModel`)
- This is a general one, but: establish type handling conventions, e.g. should `OperatorModel.evaluate` return an `Operator`? Should we have `evaluate` and `_evaluate`, with `evaluate` returning an `Operator` and `_evaluate` returning an array?

chriseclectic reviewed (approved):
Lets merge this and continue with another PR
